compiler/optimizer/PrefetchInsertion.cpp

```diff
@@ -433,8 +452,10 @@ void TR_PrefetchInsertion::examineNode(TR::TreeTop *treeTop, TR::Block *block, T
             int64_t stepInBytes = mulConstBytes * (mulConst * (int64_t)biv->getDeltaOnBackEdge() + addConst);
             TR_Structure *loop1= treeTop->getEnclosingBlock()->getStructureOf()->getContainingLoop();
             bool isTreetopInLoop = (loop1 && (loop1->asRegion() == loop)) ? true : false;
-            if (isTreetopInLoop && ((stepInBytes > 0 &&  stepInBytes <= TR::Compiler->vm.heapTailPaddingSizeInBytes()) ||
-                (stepInBytes < 0 && -stepInBytes <= TR::Compiler->om.contiguousArrayHeaderSizeInBytes())))
+
+            if (isTreetopInLoop && (
+               (stepInBytes > 0 &&  stepInBytes <= TR::Compiler->vm.heapTailPaddingSizeInBytes()) ||
+               (stepInBytes < 0 && -stepInBytes <= TR::Compiler->om.contiguousArrayHeaderSizeInBytes())))
```

Comment:
@ymanton you were right, prefetch insertion is disabled for forward arrays because `heapTailPaddingSizeInBytes` API returns 0, so for forward arrays the above `if` will never be true.

This seems quite restrictive. Tagging @dmitripivkine and @amicic here. Dmitri mentioned we removed the padding past the heap unless it is -Xaggressive. Is it possible to add this back by default and have an extra 16 bytes, or whatever the minimum size of the object is past the end of the heap? This will allow us to enable optimizations such as prefetch insertion for forward loops which are by far more common.

Reply:
> This seems quite restrictive. Tagging @dmitripivkine and @amicic here. Dmitri mentioned we removed the padding past the heap unless it is -Xaggressive. Is it possible to add this back by default and have an extra 16 bytes, or whatever the minimum size of the object is past the end of the heap? This will allow us to enable optimizations such as prefetch insertion for forward loops which are by far more common.

I would prefer not to do it. This is absolutely devastating concept caused a lot of architectural problems over the years. We worked hard to remove it and I am sure it is not paid back.